### ISSUE TITLE
HIGH: [Reliability] When opening a new DM to an existing user, draft message is lost

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#38983]
Problem:
====
When you open a DM for the first time to an existing user, we create an offline/optimistic DM -- and then swap the real one in.  In the process, however, whatever draft you were optimistically writing to the user gets blown away. To reproduce:

1. Alice and Bob are existing users
2. Alice and Bob do not have an existing DM, however
3. Alice opens a DM to Bob -- an optimistic DM is created as a placeholder
4. Alice begins typing a message in the optimistic DM, but doesn't send it
5. The server responds with the true DM to the real user
6. The draft message is lost

Solution:
====
The draft message should be retained through the switcharoo to the final correct DM.

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When opening a new DM to an existing user, draft message is lost


### What is the root cause of that problem?
When we switch to the current report, we don't update the draft common from the optimistic DM to this current DM so the draft message is lost.

https://github.com/Expensify/App/blob/d8361eee3486ccb5472060f2af7655e96461471e/src/libs/actions/Report.ts#L1100

### What changes do you think we should make in order to solve the problem?
We should get the draft comment of optimistic DM and update this to the current DM when we switch to this report here.

```
const draftComment = getDraftComment(report.reportID);
saveReportComment(report.reportID, draftComment);
```
https://github.com/Expensify/App/blob/d8361eee3486ccb5472060f2af7655e96461471e/src/libs/actions/Report.ts#L1100

OPTIONAL: we can only update the draft if the current isn't have the draft
### What alternative solutions did you explore? (Optional)
NA

**Comment by Affaq-Ahmed** [comment]
To address this issue, you can implement a mechanism to save the draft message locally on the client side when the optimistic DM is created. Then, when the real DM is received from the server, you can check if there's a saved draft message for that user and load it into the real DM if it exists. This way, the draft message won't be lost during the swap process.

**Comment by OleksandrBashchenko** [comment]
Hi, how can I contribute to this task? May I report my solutions at here? Where is your source code that I can build on my local?

**Comment by Santhosh-Sellavel** [comment]
Unable to reproduce this

https://github.com/Expensify/App/assets/85645967/6da75c32-1934-455d-b624-71a980ca4922

**Comment by quinthar** [comment]
Whoops, the initial reproduction steps were wrong; I've fixed them.  Here's an example -- note the draft comment I started writing disappeared.

https://github.com/Expensify/App/assets/575186/97265a31-0507-4a64-88f4-d4fe71554c3c

**Comment by quinthar** [comment]
@Santhosh-Sellavel can you try with the new steps?

**Comment by Santhosh-Sellavel** [comment]
Yeah was able to reproduce easily while offline or when the internet speed is low. 

https://github.com/Expensify/App/assets/85645967/6a0403f4-3fdb-4f75-9e47-53586ade94ea

**Comment by quinthar** [comment]
Great, what's the next step?  Can we get a proposal for this?  I get hit by it _all the time_.

**Comment by miljakljajic** [comment]
@Santhosh-Sellavel what do you think of the initial proposal that @nkdengineer has shared?

@Affaq-Ahmed @OleksandrBashchenko if you have proposals to share for this issue, or questions, please let us know.

**Comment by Santhosh-Sellavel** [comment]
@nkdengineer's proposal didn't work for me. So we are waiting for proposals here

**Comment by nkdengineer** [comment]
@Santhosh-Sellavel I tested and it works well. Actually when we search for new user `SearchForReport` API return the `login` of user so we still see this user in start chat. You can try to logout and login again after start chat with new user and continue the test steps.

https://github.com/Expensify/App/assets/161821005/e7e997a5-c61f-43da-be46-a0f2abb92f11

**Comment by nkdengineer** [comment]
@Santhosh-Sellavel The test branch here https://github.com/nkdengineer/App/tree/fix/38983. If it still doesn't work, please share the video here so I can see what is the problem thanks.

**Comment by Santhosh-Sellavel** [comment]
@nkdengineer Tried again still didn't work
https://github.com/Expensify/App/assets/85645967/6c88974e-8e52-4361-bd7a-d2192a5887ef

**Comment by nkdengineer** [comment]
@Santhosh-Sellavel the problem is here because, we save the draft with debounced 1s so if the `OpenReport` complete quickly the draft of the optimistic report isn't save and then we can't see this after switching to the current report. You can test with offline mode of slow 3G network, you will not see this problem.

I suggest we can decrease the debounce time from `1000` to `500` or smaller. So the time is enough to save the draft before we switch to the current report.

https://github.com/Expensify/App/blob/a1f9cecb6c81a6d2728905c67c39d33888c72b82/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.tsx#L335-L340

**Comment by Santhosh-Sellavel** [comment]
We can't go on an assumption that it would work on 500 ms, and come with a solid approach. We should save the draft before this report change happens, with some hook if possible

**Comment by quinthar** [comment]
Yes, race conditions are bugs, not solutions.  Let's make a real solution.

**Comment by Santhosh-Sellavel** [comment]
Will review shortly

**Comment by Santhosh-Sellavel** [comment]
@nkdengineer Still not working can you please check and update thanks!

**Comment by nkdengineer** [comment]
@Santhosh-Sellavel I updated the test branch https://github.com/nkdengineer/App/tree/fix/38983, please help to check again.

**Comment by Santhosh-Sellavel** [comment]
@nkdengineer Still not working

https://github.com/Expensify/App/assets/85645967/3f0a097d-336f-442d-a370-c6d5cc15bd82

**Comment by nkdengineer** [comment]
Thanks for your feedback, I will update my solution soon.

**Comment by quinthar** [comment]
@nkdengineer What is your ETA for the updated solution?

**Comment by nkdengineer** [comment]
@quinthar Sorry for delay, I will give an update today.

**Comment by nkdengineer** [comment]
> @nkdengineer Still not working

Thanks for your feedback, the problem is when the current report is already mounted, the time to save the draft comment is more than the time we switch to the current report so this doesn't work.

Updated proposal https://github.com/Expensify/App/issues/38983#issuecomment-2019570054 and test branch https://github.com/nkdengineer/App/tree/fix/38983 by passing the callback to save draft function and only call this after the draft is saved in Onyx.

**Comment by Santhosh-Sellavel** [comment]
@nkdengineer I've checked out your branch and tested it, but it's still not working for me

**Comment by Santhosh-Sellavel** [comment]
We can double this to get more eyes
cc: @miljakljajic

**Comment by ahmedGaber93** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When opening a new DM to an existing user, draft message is lost

### What is the root cause of that problem?
When opening a DM for the first time to an existing user, an offline/optimistic DM is creating, then swap to the real one when back to online/API response.
The issue here is when we typing before the offline/optimistic DM swapped, the draft comment is saved to optimistic report, then when we back online/API response, the real report not have this draft comment.

### What changes do you think we should make in order to solve the problem?
We need 2 steps to fix this.
1. get the draftComment related to offline/optimistic DM and save it to preexistingReport [here](https://github.com/Expensify/App/blob/db9e01cc302df7f2b3dd53a99c1082404676aedb/src/libs/actions/Report.ts#L1139).
```javascript
const draftComment = getDraftComment(report.reportID);
saveReportDraftComment(report.preexistingReportID, draftComment ?? '');
```

2. the step 1 cause race condition, because calling `saveReportDraftComment` in step 1 not immediately change `draftCommentCollection` as we use `waitForCollectionCallback: true` which work as promise, so `getDraftComment` in `ComposerWithSuggestions` not return the updated value in the first time, then `value` state default value not set correctly.
https://github.com/Expensify/App/blob/db9e01cc302df7f2b3dd53a99c1082404676aedb/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.tsx#L272-L278 
https://github.com/Expensify/App/blob/db9e01cc302df7f2b3dd53a99c1082404676aedb/src/libs/DraftCommentUtils.ts#L11

so the step 2 should be removing `waitForCollectionCallback` and change `Onyx.connect` to be 
```javascript
// callback: (nextVal) => {
callback: (value, key) => {
    if (key === null){
        return;
    }
    draftCommentCollection[key] = value;
},
// waitForCollectionCallback: true,
```

And removing `waitForCollectionCallback` will not affect because we check null value in [isValidDraftComment](https://github.com/Expensify/App/blob/db9e01cc302df7f2b3dd53a99c1082404676aedb/src/libs/DraftCommentUtils.ts#L27) 
### What alternative solutions did you explore? (Optional)
we can keep `waitForCollectionCallback` and create a method to update `draftCommentCollection` directly
```
draftCommentCollection[real report id] = draft message;
```  
or create new `useEffect` in `ComposerWithSuggestions` to update the `value` state when `draftComment` change and `value` not equal `draftComment`, and also we can

**Comment by miljakljajic** [comment]
@Santhosh-Sellavel what do you think of the latest proposal?

**Comment by quinthar** [comment]
@Santhosh-Sellavel When will you be able to review and respond?

**Comment by Santhosh-Sellavel** [comment]
@ahmedGaber93's proposal seems to be working on the offline case. I noticed a weird behavior, after the report change composer focus is set at the beginning of the comment. This affects continuity while typing there.

But online even weirder sometimes partially texts get lost, not working at all

https://github.com/Expensify/App/assets/85645967/60c6c271-4ed3-4787-ab13-9f2b0d5034c3

**Comment by nkdengineer** [comment]
@Santhosh-Sellavel Can you please share the case that doesn't work from my proposal, I tested it locally and it works for both offline and online case.

**Comment by Santhosh-Sellavel** [comment]
Not sure what went wrong, last time. But I run into the following issue in your proposal as well

>  I noticed a weird behavior, after the report change composer focus is set at the beginning of the comment.

Also, why would we need to pass prop`preexistingReportID` to the `ComposerWithSuggestions` component?

**Comment by ahmedGaber93** [comment]
> I noticed a weird behavior, after the report change composer focus is set at the beginning of the comment. This affects continuity while typing there.

@Santhosh-Sellavel This is the current behavior now in the app, I am not sure if we have an issue for this or not.



https://github.com/Expensify/App/assets/41129870/97514c24-3f06-49f9-bbd0-c3a487243959

**Comment by ahmedGaber93** [comment]
> But online even weirder sometimes partially texts get lost, not working at all.

@Santhosh-Sellavel This because `debouncedSaveReportComment` debounce every 1000 mills, as I remember this function is build for performance purpose. 
I will try to look for alternatives to reduce debounce time, but we can try to reduce in only when report is optimistic.

**Comment by quinthar** [comment]
@Santhosh-Sellavel Can you give a ETA for next steps?

**Comment by nkdengineer** [comment]
> Also, why would we need to pass proppreexistingReportID to the ComposerWithSuggestions component?

@Santhosh-Sellavel Sorry for the confuse, it's just a unnecessary code.


Just tested with my test branch, it works for both offline and online cases

1. Offline case


https://github.com/Expensify/App/assets/161821005/1033518d-12fb-4c27-9050-2d63fb63f055

2. Online case


https://github.com/Expensify/App/assets/161821005/dc7444e9-5be0-4a11-8db5-10053a5718d0

**Comment by ahmedGaber93** [comment]
@Santhosh-Sellavel, my proposal also works fine in the cases in the above videos from the other contributors for both online/offline and code changes is very simple.

and for fast typing cases, see my the updates here https://github.com/Expensify/App/issues/38983#issuecomment-2060295624

**Comment by Santhosh-Sellavel** [comment]
@nkdengineer I just performed steps a bit differently, instead of waiting for the network I switched to the other chat. when the network is back tried to navigate to the chat but couldn't find the chat in LHN and the draft message is lost here. The core problem is still unsolved

https://github.com/Expensify/App/assets/85645967/2792ff93-b8a7-4b08-84a6-5aa732c70692

The same case fails for @ahmedGaber93's proposal too!

**Comment by Santhosh-Sellavel** [comment]
https://github.com/Expensify/App/assets/85645967/f7d7228c-f7c9-4a2b-9e49-c6f47afce146


Even if we wait for the existing DM to load, type the message then after switching to a different chat, the draft message gets lost. We are missing something else in context here, need a clear root cause & foolproof solution.

cc: @quinthar

**Comment by quinthar** [comment]
@Santhosh-Sellavel I don't follow your scenario.  Can you present it as a numbered series of explicit steps, with each button press being a separate number?

Regardless, it sounds like you are describing a different scenario than what's in the [original post](https://github.com/Expensify/App/issues/38983#issue-2207425443).

**Comment by quinthar** [comment]
The point is to allow you to:

1. Open a chat to someone you've never spoken to *on this client* (but *have* on a different client) -- this will create an optimistic DM
2. Begin typing a message into the optimistic DM
3. The client will realize that actually there's an existing DM, and will open it
4. The partially-entered message should remain in the composer, such that you can just keep typing without realizing that the report was replaced underneath.
5.

**Comment by Santhosh-Sellavel** [comment]
1. Alice and Bob are existing users
2. Alice and Bob do have an existing DM, however but have not messaged yet since Alice opened the client
3. Alice opens a DM to Bob -- an optimistic DM is created as a placeholder
4. Alice begins typing a message in the optimistic DM, but doesn't send it (Draft 1)
5. Switches to another chat before server responds with the true DM
6. Alice opens a DM to Bob -- an optimistic DM is created as a placeholder (a new optimistic DM without Draft 1)
7. Alice begins typing a message in the optimistic DM, but doesn't send it (Draft 2)
8. The server responds with the true DM to the real user.

@quinthar These are the steps, a new optimistic DM is created every time when we open tBob's chat until we send a message on this client

**Comment by nkdengineer** [comment]
> @nkdengineer I just performed steps a bit differently, instead of waiting for the network I switched to the other chat. when the network is back tried to navigate to the chat but couldn't find the chat in LHN and the draft message is lost here. The core problem is still unsolved

@Santhosh-Sellavel  The reason is I'm only emit the event if we switch to the current report. So in the case we open another chat before online, it doesn't work.

https://github.com/Expensify/App/blob/d8361eee3486ccb5472060f2af7655e96461471e/src/libs/actions/Report.ts#L1104

Updated proposal https://github.com/Expensify/App/issues/38983#issuecomment-2019570054 to fix this

**Comment by Santhosh-Sellavel** [comment]
@nkdengineer's proposal seems to be working well. 

We have a scenario [here](https://github.com/Expensify/App/issues/38983#issuecomment-2069715821) were we might end up with two drafts. What should we do here which draft should we preserve?

**Comment by quinthar** [comment]
I think take the most recent draft.

**Comment by Santhosh-Sellavel** [comment]
Cool, then let's go with @nkdengineer's [proposal](https://github.com/Expensify/App/issues/38983#issuecomment-2019570054). LGTM.

Adding for CME's Review
C+ Reviewed
🎀 👀 🎀

**Comment by bondydaa** [comment]
not overdue just assigned out

**Comment by nkdengineer** [comment]
@Santhosh-Sellavel The PR is [here](https://github.com/Expensify/App/pull/41338).

**Comment by miljakljajic** [comment]
@Santhosh-Sellavel just confirming you are paid via manual requests?

**Comment by Santhosh-Sellavel** [comment]
Yes via manual requests

**Comment by miljakljajic** [comment]
Great! Offer is sent to @nkdengineer and I'll be ready to pay on the 20th. Thank you.

**Comment by miljakljajic** [comment]
@nkdengineer has been paid in upwork.

Payment sumary:

@Santhosh-Sellavel is owed 500 USD for their work reviewing the PR attached to this issue.

**Comment by JmillsExpensify** [comment]
$500 approved for @Santhosh-Sellavel
